File: android/app/src/main/java/dev/yury/mathmark/Merge.kt
package dev.yury.mathmark

/**
 * Сведение двух версий одного файла при синхронизации.
 *
 * Правило: если один и тот же пункт отмечен по-разному на телефоне и на
 * компьютере, выигрывает более продвинутое состояние — пройдено бьёт
 * «разбираю», «разбираю» бьёт пустое. Вперёд, а не назад.
 *
 * Если разошёлся сам текст, а не отметки, программа ничего не выбирает:
 * сообщает о споре и оставляет своё, а чужое сохраняется отдельным файлом.
 *
 * Точный повтор `merge.py` из настольной версии, вплоть до тех же тестов.
 */
object Merge {

    data class Result(val text: String, val conflict: Boolean)

    private fun rank(m: Mark): Int = when (m) {
        Mark.NONE -> 0
        Mark.HALF -> 1
        Mark.DONE -> 2
    }

    /**
     * @param base то, что было при прошлой синхронизации; null, если файл
     *             появился сразу с двух сторон.
     */
    fun merge(base: String?, local: String, remote: String): Result {
        if (local == remote) return Result(local, false)
        if (base != null && local == base) return Result(remote, false)
        if (base != null && remote == base) return Result(local, false)

        val ll = local.split("\n")
        val rl = remote.split("\n")
        if (ll.size != rl.size) return Result(local, true)

        val out = ArrayList<String>(ll.size)
        for (i in ll.indices) {
            val a = ll[i]
            val b = rl[i]
            if (a == b) {
                out.add(a)
                continue
            }
            val sa = MdItems.stripMark(a)
            val sb = MdItems.stripMark(b)
            if (sa == null || sb == null || sa != sb) {
                return Result(local, true)      // разошёлся текст, а не отметка
            }
            val ma = MdItems.markOf(a) ?: return Result(local, true)
            val mb = MdItems.markOf(b) ?: return Result(local, true)
            out.add(if (rank(ma) >= rank(mb)) a else b)
        }
        return Result(out.joinToString("\n"), false)
    }
}

File: desktop/mathmark/merge.py
"""
Сведение двух версий одного файла при синхронизации.

Правило, о котором договорились: если один и тот же пункт отмечен по-разному
на телефоне и на компьютере, **выигрывает более продвинутое состояние** —
пройдено бьёт «разбираю», «разбираю» бьёт пустое. Вперёд, а не назад.

Всё остальное — разные строки, разные файлы — сходится само, потому что
отметка меняет один символ и никогда не двигает строки.

Если же разошёлся сам текст, а не отметки, программа ничего не выбирает:
говорит о споре и отдаёт обе версии наверх.
"""

from __future__ import annotations

from .md_items import ITEM, Mark, _pairs

RANK = {Mark.NONE: 0, Mark.HALF: 1, Mark.DONE: 2}


def _strip_mark(line: str) -> str | None:
    """Строка без символа отметки — чтобы сравнивать «всё, кроме галочки»."""
    m = ITEM.fullmatch(line)
    if not m or not _pairs(m.group(2), m.group(4)):
        return None
    return m.group(1) + "- " + m.group(2) + m.group(4) + " " + m.group(5)


def merge(base: str | None, local: str, remote: str) -> tuple[str, bool]:
    """
    Свести версии. Возвращает текст и признак спора.

    base — то, что было при прошлой синхронизации; None, если файл появился
    сразу с двух сторон.
    """
    if local == remote:
        return local, False
    if base is not None and local == base:
        return remote, False
    if base is not None and remote == base:
        return local, False

    ll = local.split("\n")
    rl = remote.split("\n")
    if len(ll) != len(rl):
        return local, True

    out: list[str] = []
    for a, b in zip(ll, rl):
        if a == b:
            out.append(a)
            continue
        sa, sb = _strip_mark(a), _strip_mark(b)
        if sa is None or sb is None or sa != sb:
            return local, True          # разошёлся текст, а не отметка
        ma = Mark.of(ITEM.fullmatch(a).group(3))
        mb = Mark.of(ITEM.fullmatch(b).group(3))
        winner = a if RANK[ma] >= RANK[mb] else b
        out.append(winner)
    return "\n".join(out), False
